eth/downloader/queue.go

```diff
@@ -529,8 +529,10 @@ func (q *queue) reserveHeaders(p *peerConnection, count int, taskPool map[common
 		}
 		// Only required if the reserve is for a body type
 		if kind == bodyType {
+			q.resultCache.lock.Lock()
```

Comment:
Hi @palango, while this does fix the data race that we were seeing, I don't think it's a good approach because I think it's a misuse of `resultCache.lock`. I believe `resultCache.lock` is intended to protect fields of result cache from concurrent use, but here it is being used to protect a `fetchResult` that just happens to be accessed in the resultCache in a function also protected by that lock. `fetchResult` is a type that is built for concurrent access and so I think a more appropriate solution would be what you suggested in your comment.

However, I think that maybe we don't even need to add a `SetBodyUndone` because the celo chain always needs to download the block body for any header it downloads because the block contains some extra fields added by us that IIUC are required for both fast sync and lightest sync (and of course full sync).

https://github.com/celo-org/celo-blockchain/blob/2e37ebe19fcc83a43b067f21cbacaf385156ab13/core/types/block.go#L207-L208

So it may be sufficient to simply ensure that all `fetchResults` are created with the body set as pending, by removing the check in `newFetchResult`. The check for empty body there just checks to see if there are txs in the block, it doesn't take into account randomness or epochSnarkData. Alternatively we could just have `Header.EmptyBody` always return false.

https://github.com/celo-org/celo-blockchain/blob/2e37ebe19fcc83a43b067f21cbacaf385156ab13/eth/downloader/queue.go#L78-L80

I'm not sure why @gastonponti chose the approach that he did here, so lets wait for his input before choosing an approach.

Reply:
> Hi @palango, while this does fix the data race that we were seeing, I don't think it's a good approach because I think it's a misuse of `resultCache.lock`. I believe `resultCache.lock` is intended to protect fields of result cache from concurrent use, but here it is being used to protect a `fetchResult` that just happens to be accessed in the resultCache in a function also protected by that lock.

I agree that the current addition of locks isn't nice, but I think the above is incorrect. The `resultStore` own the `fetchResult`s until they are pulled via `GetCompleted`. In this case, `AddFetch` is used to create a new item in the store, so access should use the same lock.

> `fetchResult` is a type that is built for concurrent access and so I think a more appropriate solution would be what you suggested in your comment.

Wouldn't it need it's own lock in that case?


Anyhow, I like the idea you propose. Let's what @gastonponti thinks.

Reply:
Ok, thanks @piersy to point me that I was checking another line, haha.
Yes, I added that line to force to download every block because we need it for the randomness (probably a bad call) 
The actual problem that we have is that `header.EmptyBody`, is returning true. We should change that. Maybe use the Extra field from the Header to send a valid response

Reply:
So what I can think of here is that we add a method to `types.Header` to return the deserialized extra data. And then we can call that in `EmptyBody` to see if the header is an `istanbul` header,  if its an istanbul header then we should always return `false` for `EmptyBody`.

But this could turn out to be a bit of a performance bottleneck and so I'd suggest actually caching the deserialized results on the header in a private field, and at that point we should probably move to always using that method when we want to access the extra data.